lib/src/embedded/reusable_isolate.dart

```diff
@@ -53,58 +48,20 @@ class ReusableIsolate {
     return ReusableIsolate._(isolate, mailbox, receivePort);
   }
 
-  /// Checks out this isolate and returns a stream of messages from it.
-  ///
-  /// This isolate is considered "checked out" until the returned stream
-  /// completes. While checked out, messages may be sent to the isolate using
-  /// [send].
-  ///
-  /// Throws a [StateError] if this is called while the isolate is already
-  /// checked out.
-  Stream<Uint8List> checkOut() {
-    if (_checkedOut) {
-      throw StateError(
-          "Can't call ResuableIsolate.checkOut until the previous stream has "
-          "completed.");
-    }
-    _checkedOut = true;
-
-    var controller = StreamController<Uint8List>(sync: true);
-
-    _subscription.onData((message) {
-      var fullBuffer = message as Uint8List;
-
-      // The first byte of messages from isolates indicates whether the entire
-      // compilation is finished (1) or if it encountered an error (2). Sending
-      // this as part of the message buffer rather than a separate message
-      // avoids a race condition where the host might send a new compilation
-      // request with the same ID as one that just finished before the
-      // [IsolateDispatcher] receives word that the isolate with that ID is
-      // done. See sass/dart-sass#2004.
-      var category = fullBuffer[0];
-      var packet = Uint8List.sublistView(fullBuffer, 1);
-
-      if (category == 2) {
-        // Parse out the compilation ID and surface the [ProtocolError] as an
-        // error. This allows the [IsolateDispatcher] to notice that an error
-        // has occurred and close out the underlying channel.
-        var (_, buffer) = parsePacket(packet);
-        controller.addError(OutboundMessage.fromBuffer(buffer).error);
-        return;
-      }
+  /// Whether this isolate is in use
+  bool get borrowed => _resource != null;
 
-      controller.sink.add(packet);
-      if (category == 1) {
-        _checkedOut = false;
-        _subscription.onData(_defaultOnData);
-        _subscription.onError(null);
-        controller.close();
-      }
-    });
-
-    _subscription.onError(controller.addError);
+  /// Request this isolate as part of a pool and mark it as in use.
+  void borrow(PoolResource resource) {
+    assert(!borrowed, 'ReusableIsolate has already been borrowed.');
```

Comment:
Checks for "did you call this at the correct time" should throw `StateError`s, not just run assertions.